File: ZeroOneMatrix.py
from typing import List
from collections import deque

# DP Version, bottom up
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        update_matrix = [[10000 + 1 for j in range(len(matrix[i]))] for i in range(len(matrix))]
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                if matrix[i][j] == 0:
                    update_matrix[i][j] = 0
        for i in range(len(matrix)): # up, left
            for j in range(len(matrix[i])):
                if update_matrix[i][j] == 0:
                    continue
                if i - 1 >= 0:
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i-1][j])
                if j - 1 >= 0:
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i][j-1])
        for i in range(len(matrix) - 1, -1, -1): # down, left
            for j in range(len(matrix[i])):
                if update_matrix[i][j] == 0:
                    continue
                if i + 1 < len(matrix):
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i+1][j])
                if j - 1 >= 0:
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i][j-1])
        for i in range(len(matrix)): # up, right
            for j in range(len(matrix[i]) - 1, -1, -1):
                if update_matrix[i][j] == 0:
                    continue
                if i - 1 >= 0:
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i-1][j])
                if j + 1 < len(matrix[i]):
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i][j+1])
        for i in range(len(matrix) - 1, -1, -1): # down, right
            for j in range(len(matrix[i]) - 1, -1, -1):
                if update_matrix[i][j] == 0:
                    continue
                if i + 1 < len(matrix):
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i+1][j])
                if j + 1 < len(matrix[i]):
                    update_matrix[i][j] = min(update_matrix[i][j], 1 + update_matrix[i][j+1])
        return update_matrix



# BFS
# class Solution:
#     def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
#         # queue = []
#         update_matrix = [[-1 for j in range(len(matrix[i]))] for i in range(len(matrix))]
#         zero_index = []
#         for i in range(len(matrix)):
#             for j in range(len(matrix[i])):
#                 if matrix[i][j] == 0:
#                     # queue.append((i, j))
#                     zero_index.append((i, j))
#                     update_matrix[i][j] = 0
#         queue = deque(zero_index)
#         while len(queue) > 0:
#             # current_row, current_col = queue.pop(0)
#             current_row, current_col = queue.popleft()
#             for m, n in zip([-1, 1, 0, 0], [0, 0, -1, 1]):
#                 neighbor_row = current_row + m
#                 neighbor_col = current_col + n
#                 if neighbor_row < 0 or neighbor_row >= len(matrix) or neighbor_col < 0 or neighbor_col >= len(
#                         matrix[neighbor_row]):
#                     continue
#                 if update_matrix[neighbor_row][neighbor_col] == -1:
#                     queue.append((neighbor_row, neighbor_col))
#                     update_matrix[neighbor_row][neighbor_col] = update_matrix[current_row][current_col] + 1
#         return update_matrix


# DFS, Wrong, TLE, recrusion depth exceeds
# class Solution:
#     def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
#         update_matrix = [[-1 for j in range(len(matrix[i]))] for i in range(len(matrix))]
#
#         def dfs(i, j):
#             if update_matrix[i][j] != -1:
#                 return update_matrix[i][j]
#             if matrix[i][j] == 0:
#                 update_matrix[i][j] = 0
#                 return update_matrix[i][j]
#             # min_value = 10000 + 1
#             neighbor_values = []
#             for m, n in zip([-1, 1, 0, 0], [0, 0, -1, 1]):
#                 neighbor_i = i + m
#                 neighbor_j = j + n
#                 if neighbor_i < 0 or neighbor_i >= len(matrix) or neighbor_j < 0 or neighbor_j >= len(
#                         matrix[neighbor_i]):
#                     continue
#                 # neighbor_value = dfs(neighbor_i, neighbor_j) if update_matrix[neighbor_i][neighbor_j] == -1 else \
#                 #     update_matrix[neighbor_i][neighbor_j]
#                 neighbor_value = dfs(neighbor_i, neighbor_j)
#                 neighbor_values.append(neighbor_value + 1)
#                 if neighbor_value == 0:
#                     break
#             update_matrix[i][j] = min(neighbor_values)
#             return update_matrix[i][j]
#
#         for i in range(len(matrix)):
#             for j in range(len(matrix[i])):
#                 if update_matrix[i][j] == -1:
#                     dfs(i, j)
#         return update_matrix


if __name__ == "__main__":
    matrix = [[0, 0, 0],
              [0, 1, 0],
              [1, 1, 1]]
    solution = Solution()
    update_matrix = solution.updateMatrix(matrix)
    print(update_matrix),

File: ZeroOneMatrix.cpp
//
// Created by Golfer on 2020/8/1.
//

#include <vector>
#include <queue>
//#include <set>
#include <cstdio>
using namespace std;

// BFS Version
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
        queue<pair<int, int>> q;
//        vector<vector<int>> update_matrix(matrix.size(), vector<int>(matrix[0].size(), -1));
        vector<vector<int>> update_matrix;
        update_matrix.resize(matrix.size());
        for (int i = 0; i < matrix.size(); i++) {
            update_matrix[i].resize(matrix[i].size());
            for (int j = 0; j < matrix[i].size(); j++) {
                if (matrix[i][j] == 0) {
                    q.push(make_pair(i, j));
                    update_matrix[i][j] = 0;
                }
                else
                    update_matrix[i][j] = -1;
            }
        }
        vector<int> row_shift = {-1, 1, 0, 0};
        vector<int> col_shift = {0, 0, -1, 1};
        while (! q.empty()) {
            pair<int, int> current = q.front();
            q.pop();
            int current_row = current.first;
            int current_col = current.second;
            for (int i = 0; i < row_shift.size(); i++) {
                int neighbor_row = current_row + row_shift[i];
                int neighbor_col = current_col + col_shift[i];
                if (neighbor_row < 0 || neighbor_row >= matrix.size() || neighbor_col < 0 || neighbor_col >= matrix[neighbor_row].size())
                    continue;
                if (update_matrix[neighbor_row][neighbor_col] == -1) {
                    update_matrix[neighbor_row][neighbor_col] = update_matrix[current_row][current_col] + 1;
                    q.push(make_pair(neighbor_row, neighbor_col));
                }
            }
        }
        return update_matrix;
    }
};

// DFS Version,
// AddressSanitizer: stack-overflow
//#define inf 10000 + 1
//
//class Solution {
//public:
//    vector<vector<int>> update_matrix;
//    vector<int> row_shift = {-1, 1, 0, 0};
//    vector<int> col_shift = {0, 0, -1, 1};
//    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
//        update_matrix.resize(matrix.size());
//        for (int i = 0; i < matrix.size(); i++) {
//            update_matrix[i].resize(matrix[i].size());
//            for (int j = 0; j < matrix[i].size(); j++) {
//                update_matrix[i][j] = inf;
//            }
//        }
//        for (int i = 0; i < matrix.size(); i++) {
//            for (int j = 0; j < matrix[i].size(); j++) {
//                if (update_matrix[i][j] == inf) {
//                    dfs(matrix, i, j);
//                }
//            }
//        }
//        return update_matrix;
//    }
//    int dfs(vector<vector<int>>& matrix, int row, int col) {
//        if (update_matrix[row][col] != inf)
//            return update_matrix[row][col];
//        if (matrix[row][col] == 0) {
//            update_matrix[row][col] = 0;
//            return update_matrix[row][col];
//        }
//        for (int i = 0; i < row_shift.size(); i++) {
//            int nei_row = row + row_shift[i];
//            int nei_col = col + col_shift[i];
//            if (nei_row < 0 || nei_row >= matrix.size() || nei_col < 0 || nei_col >= matrix[nei_row].size())
//                continue;
//            update_matrix[row][col] = min(update_matrix[row][col], dfs(matrix, nei_row, nei_col) + 1);
//        }
//        return update_matrix[row][col];
//    }
//};

//int main() {
//
//}

